src/hypothesis/strategies.py

```diff
@@ -1005,39 +1005,91 @@ def fractions(min_value=None, max_value=None, max_denominator=None):
     check_valid_interval(min_value, max_value, 'min_value', 'max_value')
-
     check_valid_integer(max_denominator)
-    if max_denominator is not None and max_denominator < 1:
-        raise InvalidArgument(
-            u'Invalid denominator bound %s' % max_denominator
-        )
 
-    denominator_strategy = integers(min_value=1, max_value=max_denominator)
+    if max_denominator is not None:
+        if max_denominator < 1:
+            raise InvalidArgument(
+                'max_denominator=%r must be >= 1' % max_denominator)
+
+        def fraction_bounds(value):
+            """Find the best lower and upper approximation for value."""
+            # Adapted from CPython's Fraction.limit_denominator here:
+            # https://github.com/python/cpython/blob/3.6/Lib/fractions.py#L219
+            if value is None or value.denominator <= max_denominator:
+                return value, value
+            p0, q0, p1, q1 = 0, 1, 1, 0
+            n, d = value.numerator, value.denominator
+            while True:
+                a = n // d
+                q2 = q0 + a * q1
+                if q2 > max_denominator:
+                    break
+                p0, q0, p1, q1 = p1, q1, p0 + a * p1, q2
+                n, d = d, n - a * d
+            k = (max_denominator - q0) // q1
+            low, high = Fraction(p1, q1), Fraction(p0 + k * p1, q0 + k * q1)
+            assert low < value < high
+            return low, high
+
+        # Take the high approximation for min_value and low for max_value
+        bounds = (max_denominator, min_value, max_value)
+        _, min_value = fraction_bounds(min_value)
+        max_value, _ = fraction_bounds(max_value)
+
+        if None not in (min_value, max_value) and min_value > max_value:
+            raise InvalidArgument(
+                'There are no rational numbers with a denominator <= %r '
+                'between min_value=%r and max_value=%r' % bounds)
+
+    if min_value is not None and min_value == max_value:
+        return just(min_value)
 
     def dm_func(denom):
-        max_num = max_value * denom if max_value is not None else None
-        min_num = min_value * denom if min_value is not None else None
+        """Take denom, construct numerator strategy, and build fraction."""
+        # Four cases of algebra to get integer bounds and scale factor.
+        min_num, max_num = None, None
+        if max_value is None and min_value is None:
+            pass
+        elif min_value is None:
+            max_num = denom * max_value.numerator
+            denom *= max_value.denominator
+        elif max_value is None:
+            min_num = denom * min_value.numerator
+            denom *= min_value.denominator
+        else:
+            min_num = min_value.numerator * denom * max_value.denominator
+            max_num = max_value.numerator * denom * min_value.denominator
+            denom *= min_value.denominator * max_value.denominator
 
```

Comment:
Sigh. I'm sorry. I'm going to be picky and make you do yet more maths.

This (and to a lesser extent the previous cases) results in a denom value that is unnecessarily large, potentially using up to three times as many bytes as necessary. Consider for example the case where `min_value.denominator == max_value.denominator == max_denominator == denominator`. We end up using a range of `max_denominator ** 3` when a range of `denominator` would suffice.

I think the thing to do here is to use least common multiple (i.e multiply then divide by the gcd) instead of multiplication here. The worst case is the same, but the best case is much better.

gcd is [available from the standard library](https://docs.python.org/2/library/fractions.html#fractions.gcd) although for funsies [it was deprecated and moved in 3.5](https://docs.python.org/dev/library/math.html#math.gcd) so needs a compat entry.